tools/mascot/src/main/java/dk/alexandra/fresco/tools/mascot/field/FieldElement.java

```diff
@@ -0,0 +1,152 @@
+    return new FieldElement(op.apply(left.toBigInteger(), right.toBigInteger()).mod(modulus),
+        this.modulus, this.bitLength);
+  }
+
+  public FieldElement pow(int exponent) {
+    return new FieldElement(this.value.pow(exponent).mod(modulus), modulus, bitLength);
+  }
+
+  @Override
+  public FieldElement add(FieldElement other) {
+    return binaryOp(BigInteger::add, this, other);
+  }
+
+  public FieldElement subtract(FieldElement other) {
+    return binaryOp(BigInteger::subtract, this, other);
+  }
+
+  public FieldElement multiply(FieldElement other) {
+    return binaryOp(BigInteger::multiply, this, other);
+  }
+
+  public FieldElement negate() {
+    return new FieldElement(value.multiply(BigInteger.valueOf(-1)).mod(modulus), modulus,
+        bitLength);
+  }
+
+  public FieldElement modInverse() {
+    return new FieldElement(value.modInverse(modulus), modulus, bitLength);
+  }
+
+  public FieldElement sqrt() {
+    BigInteger rawSqrt = MathUtils.modularSqrt(value, modulus);
+    return new FieldElement(rawSqrt, modulus, bitLength);
+  }
+
+  public BigInteger toBigInteger() {
+    return this.value;
+  }
+
+  public boolean getBit(int bitIndex) {
+    return value.testBit(bitIndex);
+  }
+
+  public FieldElement select(boolean bit) {
+    return bit ? this : new FieldElement(BigInteger.ZERO, modulus, bitLength);
+  }
+
+  public boolean isZero() {
+    return value.equals(BigInteger.ZERO);
+  }
+
+  /**
+   * Converts value into byte array. <br>
+   * Result is guaranteed to exactly bitLength / 8 long.
+   *
+   * @return byte representation of value
+   */
+  public byte[] toByteArray() {
+    int byteLength = bitLength / 8;
+    byte[] res = new byte[byteLength];
+    byte[] array = value.toByteArray();
+    int arrayStart = array.length > byteLength ? array.length - byteLength : 0;
+    int resStart = array.length > byteLength ? 0 : byteLength - array.length;
+    int len = Math.min(byteLength, array.length);
+    System.arraycopy(array, arrayStart, res, resStart, len);
+    return res;
+  }
+
+  public StrictBitVector toBitVector() {
+    return new StrictBitVector(toByteArray());
+  }
+
+  public BigInteger getModulus() {
+    return this.modulus;
+  }
+
+  public int getBitLength() {
+    return bitLength;
+  }
+
+  public BigInteger getValue() {
+    return value;
+  }
+
+  @Override
+  public String toString() {
+    return "FieldElement [value=" + value + ", modulus=" + modulus + ", bitLength=" + bitLength
+        + "]";
+  }
+
+  private void sanityCheck(BigInteger value, BigInteger modulus, int bitLength) {
+    if (bitLength % 8 != 0) {
+      throw new IllegalArgumentException("Bit length must be multiple of 8");
+    } else if (value.signum() == -1) {
+      throw new IllegalArgumentException("Cannot have negative value");
+    } else if (modulus.signum() == -1) {
+      throw new IllegalArgumentException("Cannot have negative modulus");
+    } else if (modulus.bitLength() != bitLength) {
```

Comment:
In this we should really derive the bitlength from the modulus at construction. This way we simplify the constructor and avoid this exception all together ... win-win.

Reply:
Done.